package uk.gov.ons.census.notifysvc.utils;

import com.google.cloud.spring.pubsub.core.PubSubTemplate;
import java.util.concurrent.ExecutionException;
import org.springframework.stereotype.Component;
import tools.jackson.core.JacksonException;
import tools.jackson.databind.ObjectMapper;
import uk.gov.ons.census.notifysvc.model.dto.event.EventDTO;

@Component
public class PubSubHelper {
  private final PubSubTemplate pubSubTemplate;

  private static final ObjectMapper objectMapper = ObjectMapperFactory.objectMapper();

  public PubSubHelper(PubSubTemplate pubSubTemplate) {
    this.pubSubTemplate = pubSubTemplate;
  }

  public void publishAndConfirm(String topic, EventDTO payload) {
    try {
      pubSubTemplate.publish(topic, objectMapper.writeValueAsBytes(payload)).get();
    } catch (ExecutionException e) {
      throw new RuntimeException("Error publishing message to PubSub topic ", e);
    } catch (JacksonException e) {
      throw new RuntimeException("Error mapping event to JSON", e);
    } catch (InterruptedException e) {
      throw new RuntimeException(e);
    }
  }
}
